crates/wasm-host/src/context.rs

```diff
@@ -0,0 +1,1069 @@
+        types::{HostFutureIncomingResponse, HostIncomingResponse, OutgoingRequest},
+        WasiHttpCtx, WasiHttpView,
+    },
+};
+
+/// The PluginContext manages access to information that needs to cross the plugin sandbox boundary.
+pub struct PluginContext {
+    /// The WASI context that determines how things like stdio map to our buffers.
+    wasi_ctx: WasiCtx,
+    /// The WASI HTTP context that allows us to manage HTTP resources.
+    wasi_http: WasiHttpCtx,
+    /// The WASI table that maps handles to resources.
+    wasi_table: ResourceTable,
+    /// The standard I/O buffers used by WASI and captured for logging.
+    pub(crate) stdio: PluginStdio,
+    /// All host configuration.
+    host_config: Arc<bulwark_config::Config>,
+    /// Plugin-specific configuration. Stored as bytes and deserialized as JSON values by the SDK.
+    ///
+    /// There may be multiple instances of the same plugin with different values for this configuration
+    /// causing the plugin behavior to be different. For instance, a plugin might define a pattern-matching
+    /// algorithm in its code while reading the specific patterns to match from this configuration.
+    guest_config: Arc<serde_json::Map<String, serde_json::Value>>,
+    /// The set of permissions granted to a plugin.
+    permissions: bulwark_config::Permissions,
+    /// The Redis connection pool and its associated Lua scripts.
+    redis_info: Option<Arc<RedisInfo>>,
+}
+
+/// Wraps a Redis connection pool and a registry of predefined Lua scripts.
+pub struct RedisInfo {
+    /// The connection pool
+    pub pool: r2d2::Pool<redis::Client>,
+    /// A Lua script registry
+    pub registry: ScriptRegistry,
+}
+
+/// A registry of predefined Lua scripts for execution within Redis.
+pub struct ScriptRegistry {
+    /// Increments a Redis key's counter value if it has not yet expired.
+    ///
+    /// Uses the service's clock rather than Redis'. Uses Redis' TTL on a best-effort basis.
+    increment_rate_limit: redis::Script,
+    /// Checks a Redis key's counter value if it has not yet expired.
+    ///
+    /// Uses the service's clock rather than Redis'. Uses Redis' TTL on a best-effort basis.
+    check_rate_limit: redis::Script,
+    /// Increments a Redis key's counter value, corresponding to either success or failure, if it has not yet expired.
+    ///
+    /// Uses the service's clock rather than Redis'. Uses Redis' TTL on a best-effort basis.
+    increment_breaker: redis::Script,
+    /// Checks a Redis key's counter value, corresponding to either success or failure, if it has not yet expired.
+    ///
+    /// Uses the service's clock rather than Redis'. Uses Redis' TTL on a best-effort basis.
+    check_breaker: redis::Script,
+}
+
+impl Default for ScriptRegistry {
+    fn default() -> ScriptRegistry {
+        ScriptRegistry {
+            increment_rate_limit: redis::Script::new(
+                r#"
+                local counter_key = "bulwark:rl:" .. KEYS[1]
```

Comment:
Yes, and that's intended to be allowed as a form of inter-plugin communication other than params. However permissions are declared in configuration, not within the plugins themselves, so that's always fully controlled by the end user, not by the plugin author.

In the specific case of rate limiter expiration windows, I believe in the current implementation, the first window declaration wins and the second increment will increment the counter but won't change the expiration window. I'm not sure that follows the principle of least surprise, so I may revisit that.

But yeah, in general, permissions are intended to prevent this from happening accidentally, but without preventing it from happening intentionally. The docs should probably discuss this in depth, although I believe currently they do not.

Reply:
@coderabbitai Please file an issue to document that multiple plugins may access the same Redis keys/values if they both have been given permission to the same prefixes. The current docs on the topic are very light: https://docs.bulwark.security/ops/configuration#state